sdk/src/azure/core/az_mqtt5_rpc_client_hfsm.c

```diff
@@ -0,0 +1,841 @@
+
+  if (child_state == root)
+  {
+    parent_state = NULL;
+  }
+  else if (
+      child_state == idle || child_state == subscribing || child_state == ready
+      || child_state == faulted)
+  {
+    parent_state = root;
+  }
+  else if (child_state == publishing)
+  {
+    parent_state = ready;
+  }
+  else
+  {
+    // Unknown state.
+    az_platform_critical_error();
+    parent_state = NULL;
+  }
+
+  return parent_state;
+}
+
+static az_result root(az_event_policy* me, az_event event)
+{
+  az_result ret = AZ_OK;
+  az_mqtt5_rpc_client_policy* this_policy = (az_mqtt5_rpc_client_policy*)me;
+
+  if (_az_LOG_SHOULD_WRITE(event.type))
+  {
+    // printf("%s ", az_span_ptr(this_policy->_internal.rpc_client->_internal.command_name));
+    _az_LOG_WRITE(event.type, AZ_SPAN_FROM_STR("az_rpc_client_policy"));
+  }
+
+  switch (event.type)
+  {
+    case AZ_HFSM_EVENT_ENTRY:
+      break;
+
+    case AZ_HFSM_EVENT_ERROR:
+    {
+      _az_RETURN_IF_FAILED(_az_mqtt5_connection_api_callback(
+          this_policy->_internal.connection,
+          (az_event){ .type = AZ_HFSM_EVENT_ERROR, .data = event.data }));
+      // TODO: uncomment when this won't fail every time
+      // if (az_result_failed(az_event_policy_send_inbound_event(me, event)))
+      // {
+      //   az_platform_critical_error();
+      // }
+      break;
+    }
+
+    case AZ_HFSM_EVENT_EXIT:
+    {
+      if (_az_LOG_SHOULD_WRITE(AZ_HFSM_EVENT_EXIT))
+      {
+        _az_LOG_WRITE(AZ_HFSM_EVENT_EXIT, AZ_SPAN_FROM_STR("az_mqtt5_rpc_client_policy: PANIC!"));
+      }
+
+      az_platform_critical_error();
+      break;
+    }
+
+    case AZ_MQTT5_EVENT_PUBACK_RSP:
+    case AZ_MQTT5_EVENT_SUBACK_RSP:
+    case AZ_EVENT_MQTT5_CONNECTION_OPEN_REQ:
+    case AZ_MQTT5_EVENT_CONNECT_RSP:
+    case AZ_EVENT_MQTT5_CONNECTION_CLOSE_REQ:
+    case AZ_MQTT5_EVENT_DISCONNECT_RSP:
+    case AZ_MQTT5_EVENT_UNSUBACK_RSP:
+    case AZ_MQTT5_EVENT_RPC_CLIENT_UNSUB_REQ:
+    case AZ_MQTT5_EVENT_PUB_RECV_IND:
+      break;
+
+    default:
+      ret = AZ_HFSM_RETURN_HANDLE_BY_SUPERSTATE;
+      break;
+  }
+
+  return ret;
+}
+
+/**
+ * @brief If the incoming pub is of the correct topic, this transitions to the ready state if needed
+ * and passes the event for the ready state to handle
+ */
+AZ_INLINE az_result send_to_ready_if_topic_matches(
+    az_mqtt5_rpc_client_policy* this_policy,
+    az_event event,
+    az_event_policy_handler source_state)
+{
+  az_mqtt5_recv_data* recv_data = (az_mqtt5_recv_data*)event.data;
+
+  // Ensure pub is of the right topic
+  if (az_span_topic_matches_sub(
+          this_policy->_internal.rpc_client->_internal.subscription_topic, recv_data->topic))
+  {
+    // transition states if requested
+    if (source_state != NULL)
+    {
+      _az_RETURN_IF_FAILED(
+          _az_hfsm_transition_peer(&this_policy->_internal.rpc_client_hfsm, source_state, ready));
+      // notify application that it can now send invoke requests
```

Comment:
Should this be removed?

Reply:
These are commented out here for now until inbound events are properly handled with the HFSM. Since this is planned work, I wanted it to be easy to add these back in. Right now, these if statements would always be skipped, but when sending inbound is fixed I don't want them to automatically start sending double events.